data/pagination.go

```diff
@@ -0,0 +1,177 @@
+package data
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"net/url"
+	"strconv"
+
+	"github.com/ONSdigital/dp-frontend-models/model"
+	errs "github.com/ONSdigital/dp-frontend-search-controller/apperrors"
+	"github.com/ONSdigital/dp-frontend-search-controller/config"
+	"github.com/ONSdigital/log.go/log"
+)
+
+const noOfPagesToDisplay = 5
+
+// LimitOptions contains all available limit parameter values
+var LimitOptions = []int{
+	10,
+	25,
+	50,
+}
+
+// reviewPagination reviews page and limit values and sets limit and page values to queryParams
+func reviewPagination(ctx context.Context, cfg *config.Config, urlQuery url.Values, validatedQueryParams *SearchURLParams) error {
+	limit := getLimitFromURLQuery(ctx, cfg, urlQuery)
+	validatedQueryParams.Limit = limit
+
+	page := getPageFromURLQuery(ctx, cfg, urlQuery)
+	validatedQueryParams.CurrentPage = page
+
+	offset, err := getOffset(ctx, cfg, page, limit)
+	if err != nil {
+		log.Event(ctx, "unable to get offset", log.Error(err), log.ERROR)
+		return err
+	}
+	validatedQueryParams.Offset = offset
+
+	return nil
+}
+
+func getLimitFromURLQuery(ctx context.Context, cfg *config.Config, query url.Values) int {
+	limit, err := strconv.Atoi(query.Get("limit"))
+	if err != nil {
+		log.Event(ctx, fmt.Sprintf("unable to convert search limit to int - set to default %d", cfg.DefaultLimit), log.INFO)
+		query.Set("limit", strconv.Itoa(cfg.DefaultLimit))
+		limit = cfg.DefaultLimit
+	}
+
+	if limit < cfg.DefaultLimit {
+		query.Set("limit", strconv.Itoa(cfg.DefaultLimit))
+		limit = cfg.DefaultLimit
+	}
+
+	if limit > cfg.DefaultMaximumLimit {
+		query.Set("limit", strconv.Itoa(cfg.DefaultMaximumLimit))
+		limit = cfg.DefaultMaximumLimit
+	}
+
+	return limit
+}
+
+func getPageFromURLQuery(ctx context.Context, cfg *config.Config, query url.Values) int {
+	page, err := strconv.Atoi(query.Get("page"))
+	if err != nil {
+		log.Event(ctx, fmt.Sprintf("unable to convert search page to int - set to default %d", cfg.DefaultPage), log.INFO)
+		query.Set("page", strconv.Itoa(cfg.DefaultPage))
+		page = cfg.DefaultPage
+	}
+
+	if page < 1 {
+		log.Event(ctx, fmt.Sprintf("page number is less than default - default to page %d", cfg.DefaultPage), log.INFO)
+		query.Set("page", strconv.Itoa(cfg.DefaultPage))
+		page = cfg.DefaultPage
+	}
+
+	return page
+}
+
+func getOffset(ctx context.Context, cfg *config.Config, page int, limit int) (offset int, err error) {
+
+	offset = (page - 1) * limit
+
+	// when the offset is negative due to negative current page number or limit
+	if offset < 0 {
+		log.Event(ctx, fmt.Sprintf("offset less than 0 - defaulted to offset %d", cfg.DefaultOffset), log.INFO)
```

Comment:
This should probably a WARN log line - if we have many of these, it could help us identify someone is doing something malicious to us. If it stayed as a INFO log, more than likely we will ignore this.